File: app/src/main/java/com/hover/stax/settings/PinsViewModel.java
package com.hover.stax.settings;

import android.app.Application;
import android.content.Context;

import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.hover.stax.channels.Channel;
import com.hover.stax.database.DatabaseRepo;

import java.util.ArrayList;
import java.util.List;

import static org.koin.java.KoinJavaComponent.get;

public class PinsViewModel extends AndroidViewModel {

    private final DatabaseRepo repo = get(DatabaseRepo.class);
    private LiveData<List<Channel>> channels;
    private LiveData<Channel> channel;

    public PinsViewModel(Application application) {
        super(application);
        loadSelectedChannels();
        channel = new MutableLiveData<>();
    }

    public LiveData<List<Channel>> getSelectedChannels() {
        return channels;
    }

    private void loadSelectedChannels() {
        if (channels == null) {
            channels = new MutableLiveData<>();
        }
        channels = repo.getSelected();
    }

    public LiveData<Channel> getChannel() {
        if (channel == null) {
            channel = new MutableLiveData<>();
        }
        return channel;
    }

    public void loadChannel(int id) {
        channel = repo.getLiveChannel(id);
    }

    void setPin(int id, String pin) {
        List<Channel> allChannels = channels.getValue() != null ? channels.getValue() : new ArrayList<>();
        for (Channel channel : allChannels) {
            if (channel.id == id) {
                channel.pin = pin;
            }
        }
    }

    void savePins(Context c) {
        List<Channel> selectedChannels = channels.getValue() != null ? channels.getValue() : new ArrayList<>();
        for (Channel channel : selectedChannels) {
            if (channel.pin != null && !channel.pin.isEmpty()) {
                channel.pin = KeyStoreExecutor.createNewKey(channel.pin, c);
                repo.update(channel);
            }
        }
    }

    void savePin(Channel channel, Context c) {
        if (channel.pin != null && !channel.pin.isEmpty()) {
            channel.pin = KeyStoreExecutor.createNewKey(channel.pin, c);
            repo.update(channel);
        }
    }

    public void clearAllPins() {
        List<Channel> selectedChannels = channels.getValue() != null ? channels.getValue() : new ArrayList<>();
        for (Channel channel : selectedChannels) {
            channel.pin = null;
            repo.update(channel);
        }
    }

    public void removeAccount(Channel channel) {
        boolean channelDefaultChanged = channel.defaultAccount;
        channel.selected = false;
        channel.defaultAccount = false;
        repo.update(channel);

        if (channels.getValue() != null && channelDefaultChanged) {
            setRandomAccountAsDefault(channel.id);
        }
    }
    private void setRandomAccountAsDefault(int excludedChannelId) {
        for (Channel c : channels.getValue()) {
            if (c.id != excludedChannelId) {
                c.defaultAccount = true;
                repo.update(c);
                return;
            }
        }
    }

    public void setDefaultAccount(Channel channel) {
        if (channels.getValue() == null) return;
        for (Channel c : channels.getValue()) {
            c.defaultAccount = c.id == channel.id;
            repo.update(c);
        }
    }
}

File: app/src/main/java/com/hover/stax/settings/PinsViewModel.kt
package com.hover.stax.settings

import android.content.Context
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.hover.stax.account.Account
import com.hover.stax.channels.Channel
import com.hover.stax.database.DatabaseRepo
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class PinsViewModel(val repo: DatabaseRepo) : ViewModel() {

    var accounts: LiveData<List<Account>> = MutableLiveData()
    var account = MutableLiveData<Account>()
    val channel = MutableLiveData<Channel>()

    init {
        loadAccounts()
    }

    private fun loadAccounts() {
        accounts = repo.allAccountsLive
    }

    fun loadAccount(id: Int) {
        viewModelScope.launch(Dispatchers.IO) {
            val a = repo.getAccount(id)
            account.postValue(a)
            channel.postValue(repo.getChannel(a!!.channelId))
        }
    }

    fun savePin(channel: Channel, c: Context) {
        if (channel.pin != null && channel.pin.isNotEmpty()) {
            channel.pin = KeyStoreExecutor.createNewKey(channel.pin, c)
            repo.update(channel)
        }
    }

    fun removeAccount(account: Account) = viewModelScope.launch(Dispatchers.IO) {
        val defaultChanged = account.isDefault

        //remove the channel from selected
        if (repo.getAccounts(account.channelId).size == 1) {
            val channel = repo.getChannel(account.channelId)!!.apply {
                selected = false
                defaultAccount = false
            }
            repo.update(channel)
        }

        //delete the account
        repo.delete(account)

        //set a random one as the default
        if (!accounts.value.isNullOrEmpty() && defaultChanged)
            accounts.value?.random()?.let {
                it.isDefault = true
                repo.update(it)
            }
    }

    fun setDefaultAccount(account: Account) {
        if (!accounts.value.isNullOrEmpty()) {
            for (a in accounts.value!!) {
                a.isDefault = a.id == account.id
                repo.update(a)
            }
        }
    }

}